core/src/main/java/com/google/ar/sceneform/lights/EnvironmentLights.kt

```diff
@@ -0,0 +1,393 @@
+                        directionalLight?.color = Color(lightColor.r, lightColor.g, lightColor.b)
+
+                        // Convert from environmental hdr's relative value to lux for filament
+                        // using base intensity value.
+                        directionalLight?.intensity =
+                            directionalIntensity * intensityScale //* exposureScale
+                    }
+                }
+            }
+            else -> {}
+        }
+    }
+
+    /**
+     * Destroys the EnvironmentLights and frees all its associated resources.
+     */
+    fun destroy() {
+        baseIndirectLight?.destroy()
+        baseIndirectLight = null
+        baseReflections?.destroy()
+        baseReflections = null
+        indirectLight?.destroy()
+        indirectLight = null
+        directionalLight?.destroy()
+        directionalLight = null
+        skybox?.destroy()
+        skybox = null
+    }
+
+    companion object {
+
+        // Convert Environmental HDR's spherical harmonics to Filament spherical harmonics.
+        // This conversion is calculated to include the following:
+        //  - pre-scaling by SH basis normalization factor [shader optimization]
+        //  - sqrt(2) factor coming from keeping only the real part of the basis [shader optimization]
+        //  - 1/pi factor for the diffuse lambert BRDF [shader optimization]
+        //  - |dot(n,l)| spherical harmonics [irradiance]
+        //  - scaling for convolution of SH function by radially symmetrical SH function [irradiance]
+        //
+        // ENVIRONMENTAL_HDR_TO_FILAMENT_SH_INDEX_MAP must be applied change ordering of coeffients
+        // from Environmental HDR to filament.
+        private val SPHERICAL_HARMONICS_RECONSTRUCTION_FACTORS =
+            floatArrayOf(
+                0.282095f, -0.325735f, 0.325735f,
+                -0.325735f, 0.273137f, -0.273137f,
+                // notice index 6 & 7 swapped
+                0.078848f, -0.273137f, 0.136569f
+            )
+
+        // SH coefficients are not in the same order in Filament and Environmental HDR.
+        // SH coefficients at indices 6 and 7 are swapped between the two implementations.
+//        private val ENVIRONMENTAL_HDR_TO_FILAMENT_SH_INDEX_MAP = intArrayOf(
+//            0, 1, 2, 3, 4, 5, 7, 6, 8
+//        )
+    }
+}
+
+//TODO : Use the IBLPrefilterContext
+// IBLPrefilterContext takes a 2D texture in parameter
+// Determine how to create the 2D texture from the ARCore CubeMap
+//    val context = IBLPrefilterContext(engine)
+//    IBLPrefilterContext.EquirectangularToCubemap(context)
+//        .run(cubeMapTexture)?.let { skyboxTexture ->
+//            IBLPrefilterContext.SpecularFilter(context)
+//                .run(skyboxTexture)?.let { reflectionsTexture ->
+//                    reflections(reflectionsTexture)
+//                }
+//        }
+//    engine.destroyTexture(cubeMapTexture)
+fun Array<ArImage>.toTexture(): Texture {
```

Comment:
This looks super expensive for being called on every frame. I'm not familiar with Filament, but is there no way to create the texture once and then update the contents afterwards? I don't think ARCore will ever decide to provide a differently sized set of images, so it would be great if you could create the texture once and directly set contents afterwards.

Reply:
You can create the texture once and update the content afterwards.